core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -623,8 +623,12 @@ private[log] class Cleaner(val id: Int,
 
     val groupedSegments = groupSegmentsBySize(log.logSegments(0, endOffset), log.config.segmentSize,
       log.config.maxIndexSize, cleanable.firstUncleanableOffset)
-    for (group <- groupedSegments)
-      cleanSegments(log, group, offsetMap, currentTime, stats, transactionMetadata, legacyDeleteHorizonMs)
+    val groupIter = groupedSegments.iterator
+    while (groupIter.hasNext) {
+      val group = groupIter.next
+      val retainLastBatch = !groupIter.hasNext
+      cleanSegments(log, group, offsetMap, currentTime, stats, transactionMetadata, legacyDeleteHorizonMs, retainLastBatch)
```

Comment:
It's probably simpler to just pass `upperBoundOffset`  to `cleanInto` and check if it equals to a batch's nextOffset.

Reply:
updated